### PR TITLE
Feature/New Xcode Template using a Fastlane BuildPlan

### PR DESCRIPTION
**Pull request by DanielKainz**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

This PR replaces the current Xcode template. 
The new template has a different structure: it now uses a xcworkspace (in the test repository) containing two projects, one for the tests and one for the exercise / solution. The BuildPlans now work with fastlane instead of a xcodebuild command. 
The new template gives students the possibility to create new files in the project and after cloning the repository they are able to open the xcodeproj file in Xcode to work on the project, build it and test it on an iPhone Simulator. 

The template gets extended in a Follow-Up PR. It then should include the strategy pattern with sorting as the other templates do. The UI component should allow selecting a sorting strategy, generate a list of random dates via a simple button tap, sort them using another button and display the (un-) sorted list.

### Description

The template structure as well as the BuildPlans changed completely. The changes are described above and below.
Also I changed the name of the text field to enter a package name while creating a Swift programming exercise. Now it is called App Name, this is used for the dynamic naming of the files.

### Steps for Testing

1. You need to use Test Server 1 (this is necessary because of Swift)
2. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise and select `Swift` as programming language and `Xcode` as project type (there should be an option to type in an `App Name`) or use the following exercise: https://artemistest.ase.in.tum.de/course-management/480/programming-exercises/10033 
4. Check that the solution result is `2 of 2 passed` and the exercise result is `0 of 2 passed`
5. Clone the three repositories (for the next steps you need Xcode installed on your machine)
    5.1 In the test repository: Open the `xcworkspace` and see the two linked projects, and the `Unit` and `UI` `tests` in the test projects; also check that a `fastlane` folder is present in the repository containing the necessary configuration files (`Appfile`, `Fastfile`, `Scanfile` and a `readme`)
    5.2 In the exercise repository: Open the `xcodeproj` and see the exercise template -> solve the tasks and push your solution
    5.3 In the solution repository: Open the `xcodeproj` and check the solution 
6. See how the result of the exercise gets updated
7. Check the BuildPlans on Bamboo
    7.1 They should have two script tasks: build -> calls `fastlane build`; test -> calls `fastlane test` and `exit 0` (this is necessary due to a warning which would cause the build to fail even if it should not fail)
   7.2 The Junit Parser should read the `**/report.junit` file
   7.3 The requirements should be: `Simulator - iOS 14.5` and `fastlane`

### Review Progress

- Code Review
  - [x] Review 1
  - [ ] Review 2
- Manual Tests
  - [x] Test 1
  - [ ] Test 2

### Screenshots

Here you can see the textfield for the App Name:
App Name

This is the testing step regarding the BuildPlan Configuration. In the video it did it with the BASE BuildPlan but the SOLUTION BuildPlan shows exactly the same:
https://user-images.githubusercontent.com/82283581/134001595-aadc9819-2d9b-4ffc-8e0c-68acca5d1282.mov

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java [hunk 1]
```diff
@@ -234,9 +240,12 @@ private Stage createBuildStage(ProgrammingLanguage programmingLanguage, ProjectT
                     defaultJob.finalTasks(scaTasks.toArray(new Task[0]));
                 }
                 if (isXcodeProject) {
+                    // add a requirement to be able to run the Xcode build tasks using fastlane
+                    var requirement1 = new Requirement("system.fastlane");
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:244]:
```suggestion
                    var requirement1 = new Requirement("system.builder.fastlane.fastlane");
```

**Review by Malyuk-A — COMMENTED** [review]
Reviewed Code, looks reasonable 👍 (Even though I have very limited experience when it comes to Xcode or BuildPlans)
I left some suggestions and comments.

@DanielKainz tested the functionality with me on TS1 so I can approve the manual testing part, but sadly without testing the local Xcode functionality because I only have a Windows PC.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 2]
```diff
@@ -597,17 +597,26 @@ public void replacePlaceholders(ProgrammingExercise programmingExercise, Reposit
...
+        }
 
         Map<String, String> replacements = new HashMap<>();
 
-        if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA || programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.KOTLIN
-                || programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.SWIFT) {
+        if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA || programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.KOTLIN) {
             replacements.put("${packageName}", programmingExercise.getPackageName());
         }
+        else if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.SWIFT && programmingExercise.getProjectType() == ProjectType.PLAIN) {
```
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java]: I combined the if statements

File: src/main/resources/templates/swift/xcode/test/${appName}UITest/${appName}UITest.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        // UI tests must launch the application that they test.
+        app = XCUIApplication(bundleIdentifier: "de.tum.in.ase")
+        app.launch()
+    }
+
+    func testGreetingUI() {
+        XCTAssertTrue(app.staticTexts["Hello, world!"].exists, "The greeting was not presented correctly!")
+    }
+    
+//    func testGreetingLogic() throws {
```
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/resources/templates/swift/xcode/test/${appName}UITest/${appName}UITest.swift]: I removed it

File: src/main/resources/templates/swift/xcode/test/fastlane/App.file [hunk 4]
```diff
@@ -0,0 +1 @@
+app_identifier "de.tum.in.ase" # The bundle identifier of your app
```
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/resources/templates/swift/xcode/test/fastlane/App.file]: I added a new line

File: src/main/resources/templates/swift/xcode/test/fastlane/Scan.file [hunk 5]
```diff
@@ -0,0 +1 @@
+devices(["iPhone 12 Pro Max"])
```
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/resources/templates/swift/xcode/test/fastlane/Scan.file]: Here as well

File: src/main/resources/templates/swift/xcode/test/fastlane/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+Run UI Tests only
+### uitest
+```
+fastlane uitest
+```
+Run UI Tests only
+
+----
+
+This README.md is auto-generated and will be re-generated every time [_fastlane_](https://fastlane.tools) is run.
```
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/resources/templates/swift/xcode/test/fastlane/README.md]: No problem. I removed the line

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 7]
```diff
@@ -597,25 +597,29 @@ public void replacePlaceholders(ProgrammingExercise programmingExercise, Reposit
...
 
         if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA || programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.KOTLIN) {
             replacements.put("${packageName}", programmingExercise.getPackageName());
         }
-        else if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.SWIFT && programmingExercise.getProjectType() == ProjectType.PLAIN) {
-            replacements.put("${packageName}", programmingExercise.getPackageName());
-        }
-        else if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.SWIFT && programmingExercise.getProjectType() == ProjectType.XCODE) {
-            replacements.put("${appName}", programmingExercise.getPackageName());
+        else if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.SWIFT) {
```
Comment (Malyuk-A) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:616]: I have left you a message on Slack providing an example of how this code could be replaced by using switch cases and extracting the programmingLanguage and ProjectType to variables beforehand, therefore the code should become way slimmer

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java]: I refactored the whole part with a switch statement

File: src/main/resources/templates/swift/xcode/test/fastlane/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/resources/templates/swift/xcode/test/fastlane/README.md:41]: Misunderstanding: I re-added the line and we agreed that the file should not be deleted

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java (hunk 7 above)
Comment (DanielKainz) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java]: Thanks. I refactored it.

**Review by Malyuk-A — APPROVED** [review]
Thank you for the quick response to my comments and refactoring of the code 👍
Approve

**Review by Olimbia23 — CHANGES_REQUESTED** [review]
I tested it manually @DanielKainz Nice work 🎉 I also will have a look at the code later today. 

Everything works as described. However, I don't know how to test your last point regarding JUnit parsing and simulator requirements. I have the following feedback:

- Please make PR description more detailed.

- I would strongly recommend you to make a short video showing the testing steps and upload it in the PR description so reviewers can easily test it afterwards. 

- Please don't remove checks from the checklist.

- I noticed two UI issues:
  - in the exercise edit page, the application name text field changes the color from grey to white, when you hover over it. I think it should remain grey as it is not editable. Screenshot below.
  - In the exercise detail page the warning icon should be hidden during the build process and it should be smaller in size (similar to the "Resubmit" button). 

Screenshot 2021-09-18 at 2 03 10 PM

---

Screenshot 2021-09-18 at 1 49 00 PM

**Review by Olimbia23 — COMMENTED** [review]
I checked the code. Left one minor suggestion.

**Comment by DanielKainz** [comment]
@Olimbia23 
Thank you very much 🙂
Regarding the last point in the testing steps: I meant clicking on the BuildPlan links and checking the script tasks, the Junit Parser task and the requirements of the Default Job Configuration 
I added a video for that task so you can check it out. 

Alright. Thanks. I’ll apply your feedback in my next PRs. 

Regarding the UI:
* The color change appears for every textfield (regardless if it is editable or not); I was not sure it this is intentional or not so I did not change anything here
* I have not changed anything regarding this icon. Maybe it is displayed as a larger icon to make it more concise.
* What do you think: Should we handle the UI issues in another PR since they are not related with the functionality of my PR and we could make sure how to handle those exactly?

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 8]
```diff
@@ -352,6 +352,18 @@ public void copyResources(Resource[] resources, String prefix, String targetDire
...
+                targetFilePath = targetFilePath.replace("Fast.file", "Fastfile");
+            }
+            // special case for 'Appfile' files which would not be included in the build otherwise
+            if (targetFilePath.endsWith("App.file")) {
+                targetFilePath = targetFilePath.replace("App.file", "Appfile");
+            }
+            // special case for 'Scanfile' files which would not be included in the build otherwise
+            if (targetFilePath.endsWith("Scan.file")) {
+                targetFilePath = targetFilePath.replace("Scan.file", "Scanfile");
+            }
```
[the remark this answers is not in the document]
Comment (DanielKainz) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:366]: I also noticed that. I could check how to handle this the best way and would do it in another PR if this is okay for you.

**Review by Olimbia23 — APPROVED** [review]
Thank you for adding the screen recording @DanielKainz 
It made it very easily to test the last point that I couldn't test the first time.
For me everything worked as expected. 🎉 
It would be great to fix the issues I mentioned before in another small PR.
Approve 👍

**Review by Malyuk-A — APPROVED** [review]
Talked with @DanielKainz about the "changed" since the last review. Re-approve.

**Pull request merged**